File: 1962.remove-stones-to-minimize-the-total/solution.py
class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        N = len(piles)

        # Build a max heap-queue ... O(N) time
        hq = []
        total = 0
        for pile in piles:
            total += pile
            heapq.heappush(hq, -pile)

		# Do operations ... O(k log N) time
        for i in range(k):
            peek = heapq.heappop(hq)
            remove = int(-peek / 2)
            peek += remove
            total -= remove
            heapq.heappush(hq, peek)

        return total

File: 1962.remove-stones-to-minimize-the-total/solution.kt
import java.util.PriorityQueue
import java.util.Comparator

class Solution {
    fun minStoneSum(piles: IntArray, k: Int): Int {
        val hq = PriorityQueue<Int>(piles.size, Comparator<Int> { a, b -> b - a })

        // Build a max heap queue
        var total = 0
        for (pile in piles) {
            total += pile
            hq.add(pile)
        }

        // Do operations k times
        repeat(k) {
            var peek = hq.poll()
            val remove = peek / 2
            total -= remove
            peek -= remove
            hq.add(peek)
        }

        return total
    }
}
